tensorflow/compiler/tf2tensorrt/convert/preperation_pass.cc

```diff
@@ -0,0 +1,310 @@
+  grappler::SetVector<NodeDef*> nodes_to_simplify;
+
+  // Stop pipeline after first stage returning non-empty simplified tensor
+  // name.
+  const auto stop = [](const string& result) { return !result.empty(); };
+
+  // Run the main pipeline.
+  const grappler::GraphOptimizerContext ctx(
+      &nodes_to_preserve_, optimized_graph_, graph_properties_.get(),
+      node_map_.get(), &feed_nodes_, opt_level_);
+  const TRTPreperationContext ctx_ext(&nodes_to_simplify);
+  grappler::GraphOptimizerStagePipeline<string> pipeline(stop);
+  TF_RETURN_IF_ERROR(RunPipeline(pipeline, ctx, ctx_ext));
+
+  // Perform cast rewrites at the end.
+  if (precision_mode_ == TrtPrecisionMode::FP16) {
+    const grappler::GraphOptimizerContext ctx(
+        &nodes_to_preserve_, optimized_graph_, graph_properties_.get(),
+        node_map_.get(), &feed_nodes_, opt_level_);
+    const TRTPreperationContext ctx_ext(&nodes_to_simplify);
+    grappler::GraphOptimizerStagePipeline<string> pipeline(stop);
+    pipeline.AddStage<RewriteFP32CastsStage>(ctx, ctx_ext);
+    TF_RETURN_IF_ERROR(RunPipeline(pipeline, ctx, ctx_ext));
+  }
+
+  return Status::OK();
+}
+
+Status TRTPreperationPass::Optimize(grappler::Cluster* cluster,
+                                     const grappler::GrapplerItem& item,
+                                     GraphDef* optimized_graph) {
+  // Set up helper data structures.
+  grappler::GrapplerItem optimized_item(item);
+  optimized_graph_ = &optimized_item.graph;
+  nodes_to_preserve_ = item.NodesToPreserve();
+  fetch_nodes_known_ = !item.fetch.empty();
+  node_map_ = std::make_unique<grappler::NodeMap>(optimized_graph_);
+  for (const auto& feed : item.feed) {
+    feed_nodes_.insert(grappler::NodeName(feed.first));
+  }
+
+  // Perform topological sort on the graph in order to help DedupComputations
+  // and AddOpsRewrite to optimize larger subgraphs starting from the roots
+  // with more inputs.
+  TF_RETURN_IF_ERROR(grappler::TopologicalSort(optimized_graph_));
+  GRAPPLER_RETURN_IF_DEADLINE_EXCEEDED();
+
+  // Setup the graph properties handle.
+  graph_properties_ =
+      std::make_unique<grappler::GraphProperties>(optimized_item);
+  const bool assume_valid_feeds = opt_level_ == RewriterConfig::AGGRESSIVE;
+  const Status status =
+      graph_properties_->InferStatically(assume_valid_feeds,
+                                         /*aggressive_shape_inference=*/false,
+                                         /*include_tensor_values=*/false);
+  const bool can_use_shapes = status.ok();
+  if (!can_use_shapes) {
+    VLOG(1) << "Shape inference failed." << status.error_message();
+  }
+
+  // Perform conditioning optimization.
+  TF_RETURN_IF_ERROR(ConditionGraphConversion(can_use_shapes));
+  *optimized_graph = *optimized_graph_;
+  return Status::OK();
+}
+
+static grappler::CustomGraphOptimizerRegistrar TRTPreperationPass_Registrar(
+    []() { return new TRTPreperationPass("TensorRTConditioner"); },
```

Comment:
The name here is not consistent with what is used in trt_convert.py.
How do you test this PR?